README.md
Python-SJCL
===========

[![Travis CI](https://travis-ci.org/berlincode/sjcl.svg?branch=master&style=flat)](https://travis-ci.org/berlincode/sjcl)
[![Python versions](https://img.shields.io/pypi/pyversions/sjcl.svg)](https://pypi.python.org/pypi/sjcl/)
[![new-style BSD](https://img.shields.io/pypi/l/sjcl.svg)](https://github.com/berlincode/sjcl/blob/master/LICENSE.txt)

Decrypt and encrypt messages compatible to the "Stanford Javascript Crypto
Library (SJCL)" message format. This is a wrapper around pycrypto.

This module was created while programming and testing the encrypted
blog platform on cryptedblog.com which is based on sjcl.

Typical usage may look like this:

```python
    #!/usr/bin/env python

    from sjcl import SJCL

    cyphertext = SJCL().encrypt(b"secret message to encrypt", "shared_secret")

    print cyphertext
    print SJCL().decrypt(cyphertext, "shared_secret")
```

Public repository
-----------------

[https://github.com/berlincode/sjcl](https://github.com/berlincode/sjcl)


License
-------

Code and documentation copyright Ulf Bartel. Code is licensed under the
[new-style BSD license](./LICENSE.txt).



CHANGES.txt
v0.1.1, 2014-05-21 -- Initial public release.
v0.1.4, 2016-04-17 -- Re-init of repository after homepage changed.
v0.1.5, 2016-07-12 -- Python3 compat
v0.1.6, 2017-07-31 -- Now dependent on pycryptodome
v0.2.0, 2018-02-22 -- AES-GCM mode support 
v0.2.1, 2018-08-16 -- Fixed README.md (missing 'b' prefix for use with python3)


tests/simple.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import unittest

from sjcl import SJCL

class Test(unittest.TestCase):

    def setUp(self):
        pass

    def test_encrypt_decrypt(self):
        message = b"secret message to encrypt"
        cyphertext = SJCL().encrypt(message, "shared_secret")
        self.assertEqual(
            SJCL().decrypt(cyphertext, "shared_secret"),
            message
        )

    def test_decrypt(self):
        cyphertext = {
            'ks': 128,
            'cipher': 'aes',
            'mode': 'ccm',
            'v': 1,
            'adata': '',
            'iv': 'fR4fZKbjsZOrzDyjCYdEQw==',
            'salt': '5IiimlH8JvY=',
            'ts': 64,
            'iter': 1000,
            'ct': 'V8BYrUdurq1/Qx/EX8EBliKDKa6XB93dZ6QOFSelw77Q'
        }
        self.assertEqual(
            SJCL().decrypt(cyphertext, "shared_secret"),
            b"secret message to encrypt"
        )

    def test_gcm_mode_encrypt_decrypt(self):
        message = b"secret message to encrypt"
        cyphertext = SJCL().encrypt(message, "shared_secret", mode="gcm")

        self.assertEqual(
            SJCL().decrypt(cyphertext, "shared_secret"),
            message
        )

    def test_gcm_decrypt(self):
        cyphertext = {
            "iv": "oi51KdYi8av0PysYeCDDiw==",
            "v":1,
            "iter":10000,
            "ks":256,
            "ts":128,
            "mode":"gcm",
            "adata":"",
            "cipher":"aes",
            "salt": "C4DRJM5AH+A=",
            "ct": "KIINDGEQO63pY2mFIWpOuWgx2RAnfU0rhKU="
        }
        self.assertEqual(
            SJCL().decrypt(cyphertext, "vYHOPyQ7Q6NOfm6zkT44IE2SVv+52arqCOv7cDfMApI="),
            b"test paste"
        )

if __name__ == '__main__':
    unittest.main()


examples/backup_cryptedblog.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Example for sjcl python module:

This module downloads all content of a blog from www.cyptedblog.com via the
JSON api, decrypts the SJCL encrpted posts/comments/images and saves both, the
encrypted and decrypted data to disk.

This module may also be used as a backup solution for cryptedblog.
"""

import os
import json
import base64

import requests
from sjcl import SJCL

URL_API = '%(server)s/api/blog/v1.0'
URL_POST = URL_API + '/blog/%(blog)s/posts?offset=%(offset)d&limit=%(limit)d'
URL_COMMENT = URL_API + '/blog/%(blog)s/posts/%(post)s/comments?offset=%(offset)d&limit=%(limit)d'
URL_IMAGES = URL_API + '/images/%(image)s'


def write_file(fname_parts, content):
    """ write a file and create all needed directories """

    fname_parts = [str(part) for part in fname_parts]
    # try to create the directory
    if len(fname_parts) > 1:
        try:
            os.makedirs(os.path.join(*fname_parts[:-1]))
        except OSError:
            pass
    # write file
    fhandle = open(os.path.join(*fname_parts), "w")
    fhandle.write(content)
    fhandle.close()


class Cryptedblog(object):
    """ This module downloads all content of a blog from www.cyptedblog.com via the
    JSON api, decrypts the SJCL encrpted posts/comments/images and saves both, the
    encrypted and decrypted data to disk.

    This module may also be used as a backup solution for cryptedblog.
    """

    def __init__(self, server, blog, shared_secret, limit=5):
        self.server = server
        self.blog = blog
        self.shared_secret = shared_secret
        self.limit = limit  # how many posts/comments to fetch with one request
        self.sjcl = SJCL()

    def download_images(self, fname_parts, key_image):
        response = requests.get(URL_IMAGES % {
            "server": self.server,
            "image": key_image
        })
        data = response.json()
        # special case: images are server from blobstrore and do normally not
        # contain a "success" field !!!
        if not data.get("success", True):
            raise Exception("error image")
        write_file(
            fname_parts + ["images-encrypted.json"],
            json.dumps(data, indent=4, sort_keys=True)
        )

        images = json.loads(
            self.sjcl.decrypt(
                data,
                self.shared_secret
            )
        )
        for (counter, image) in enumerate(images):
            write_file(
                fname_parts + ["image_%02d.json" % counter],
                json.dumps(image, indent=4, sort_keys=True)
            )
            # data uri of inline images starts like "data:image/jpeg;base64,"
            file_type = image["data"].split("image/")[1].split(";")[0]
            binary_data = base64.b64decode(image["data"].split("base64,")[1])
            write_file(
                fname_parts + ["image_%02d.%s" % (counter, file_type)],
                binary_data
            )

    def download_comments(self, key_post):
        offset = 0
        while True:

            response = requests.get(URL_COMMENT % {
                "server": self.server,
                "blog": self.blog,
                "post": key_post,
                "offset": offset,
                "limit": self.limit
            })
            data = response.json()
            if not data.get("success", False):
                raise Exception("error comment")

            for comment in data["data"]:
                key_comment = comment["comment"]
                fname_parts = [self.blog, "posts", key_post, "comments", key_comment]

                # write original data
                write_file(
                    fname_parts + ["comment.json"],
                    json.dumps(comment, indent=4, sort_keys=True)
                )
                # decrypt and write again
                comment["comment-decrypted"] = json.loads(
                    self.sjcl.decrypt(
                        comment["comment-encrypted"],
                        self.shared_secret
                    )
                )
                write_file(
                    fname_parts + ["comment-decrypted.json"],
                    json.dumps(comment, indent=4, sort_keys=True)
                )
                # get images
                if "images" in comment:
                    self.download_images(fname_parts + ["images"], comment["images"])

                offset += 1

            if len(data["data"]) == 0:
                break

    def download_posts(self):
        offset = 0
        while True:
            response = requests.get(URL_POST % {
                "server": self.server,
                "blog": self.blog,
                "offset": offset,
                "limit": self.limit
            })
            data = response.json()
            for post in data["data"]:
                key_post = post["post"]
                fname_parts = [self.blog, "posts", key_post]

                # write original data
                write_file(
                    fname_parts + ["post.json"],
                    json.dumps(post, indent=4, sort_keys=True)
                )

                # decrypt and write again
                post["post-decrypted"] = json.loads(
                    self.sjcl.decrypt(
                        post["post-encrypted"],
                        self.shared_secret
                    )
                )
                write_file(
                    fname_parts + ["post_decrypted.json"],
                    json.dumps(post, indent=4, sort_keys=True)
                )

                # get images
                if "images" in post:
                    self.download_images(fname_parts + ["images"], post["images"])

                self.download_comments(key_post)
                offset += 1

            # break if there are no further posts
            if len(data["data"]) == 0:
                break

if __name__ == "__main__":
    cr = Cryptedblog(
        server="http://www.cryptedblog.com",
        blog="test-blog",
        shared_secret="YourSharedSecret",
    )
    cr.download_posts()


sjcl/sjcl.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Decrypt and encrypt messages compatible to the "Stanford Javascript Crypto
Library (SJCL)" message format.

This module was created while programming and testing the encrypted
blog platform on cryptedblog.com which is based on sjcl.
"""

import base64

from Crypto.Hash import SHA256, HMAC
from Crypto.Protocol.KDF import PBKDF2
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes

# Default tag length for differend AES modes
DEFAULT_TLEN = {
    AES.MODE_CCM: 64,
    AES.MODE_GCM: 128
}

# field explanation from Jay Tuley (jbtule)
# - see http://stackoverflow.com/a/13570154
#
#    ct: cipher-text your encrypted data, obviously
#    iv: Should be unique and not be reused with the same key for AES-CCM
#    salt: random, and is used to create the key from password with Pbkdf2
#    adata: additional authenticated data, is plain text data that you want
#        to include but ensure that it has not been tampered when using
#        AES-CCM. If you aren't ever going to include any data then you can
#        ignore it (you would have had to pass it with the plaintext in sjcl).
#
#    iter: iterations for Pbkdf2, this just needs to be high enough to slow
#        down bruteforce on your password needs to change with the speed of
#        hardware in the future.
#    ks: keysize to know what size key to generate with Pbkdf2, needs to change
#         with the amount of security in future
#    ts: tagsize to know what size authentication tag is part of your cipher
#         text
#    cipher: If you will only support AES, then you can just assume.
#    mode: if you will only support AES-CCM than you can just assume.
#    v: If you will only support one version of sjcl in the future than you can
#         just assume.


#def hex_string(array):
#    import binascii
#    return binascii.hexlify(bytearray(array))


def truncate_iv(iv, ol, tlen):  # ol and tlen in bits
    ivl = len(iv)  # iv length in bytes
    ol = (ol - tlen) // 8

    # "compute the length of the length" (see ccm.js)
    L = 2
    while (L < 4) and ((ol >> (8*L))) > 0:
        L += 1
    if L < 15 - ivl:
        L = 15 - ivl

    return iv[:(15-L)]


def get_aes_mode(mode):
    """Return pycrypto's AES mode, raise exception if not supported"""
    aes_mode_attr = "MODE_{}".format(mode.upper())
    try:
        aes_mode = getattr(AES, aes_mode_attr)
    except AttributeError:
        raise Exception(
            "Pycrypto/pycryptodome does not seem to support {}. ".format(aes_mode_attr) +
            "If you use pycrypto, you need a version >= 2.7a1 (or a special branch)."
        )
    return aes_mode


class SJCL(object):

    def __init__(self):
        self.salt_size = 8  # bytes
        self.prf = lambda p, s: HMAC.new(p, s, SHA256).digest()

    def decrypt(self, data, passphrase):
        if data["cipher"] != "aes":
            raise Exception("only aes cipher supported")

        aes_mode = get_aes_mode(data["mode"])
        tlen = data["ts"]

        if data["adata"] != "":
            raise Exception("additional authentication data not equal ''")

        if data["v"] != 1:
            raise Exception("only version 1 is currently supported")

        # Fix padding
        if aes_mode == AES.MODE_CCM:
            # not a multiple of 4, add padding:
            if len(data["salt"]) % 4:
                data["salt"] += '=' * (4 - len(data["salt"]) % 4)
        salt = base64.b64decode(data["salt"])

    #    print "salt", hex_string(salt)
        if len(salt) != self.salt_size:
            raise Exception("salt should be %d bytes long" % self.salt_size)

        dkLen = data["ks"]//8
        if dkLen != 16 and dkLen != 32:
            raise Exception("key length should be 16 bytes or 32 bytes")

        key = PBKDF2(
            passphrase,
            salt,
            count=data['iter'],
            dkLen=dkLen,
            prf=self.prf
        )
#       print "key", hex_string(key)
        if aes_mode == AES.MODE_CCM:
            # Fix padding
            if len(data["iv"]) % 4:
            # not a multiple of 4, add padding:
                data["iv"] += '=' * (4 - len(data["iv"]) % 4)
            if len(data["ct"]) % 4:
            # not a multiple of 4, add padding:
                data["ct"] += '=' * (4 - len(data["ct"]) % 4)

        ciphertext = base64.b64decode(data["ct"])
        iv = base64.b64decode(data["iv"])
#        print AES.block_size

        if aes_mode == AES.MODE_CCM:
            nonce = truncate_iv(iv, len(ciphertext)*8, data["ts"])
        else:
            nonce = iv

        # split tag from ciphertext (tag was simply appended to ciphertext)
        mac = ciphertext[-(data["ts"]//8):]
#        print len(ciphertext)
        ciphertext = ciphertext[:-(data["ts"]//8)]
#        print len(ciphertext)
#        print len(tag)

#        print "len", len(nonce)
        cipher = AES.new(key, aes_mode, nonce, mac_len=tlen // 8)
        plaintext = cipher.decrypt(ciphertext)

        cipher.verify(mac)

        return plaintext

    def encrypt(self, plaintext, passphrase, mode="ccm", count=10000, dkLen=16):
        # dkLen = key length in bytes
        aes_mode = get_aes_mode(mode)
        tlen = DEFAULT_TLEN[aes_mode]

        salt = get_random_bytes(self.salt_size)
        iv = get_random_bytes(16)  # TODO dkLen?

        key = PBKDF2(passphrase, salt, count=count, dkLen=dkLen, prf=self.prf)

        # TODO plaintext padding?
        if aes_mode == AES.MODE_CCM:
            nonce = truncate_iv(iv, len(plaintext) * 8, tlen)
        else:
            nonce = iv
    #    print len(nonce)

        cipher = AES.new(
            key,
            aes_mode,
            nonce,
            mac_len=tlen // 8
        )

        ciphertext = cipher.encrypt(plaintext)
        mac = cipher.digest()

        ciphertext = ciphertext + mac

        return {
            "salt": base64.b64encode(salt),
            "iter": count,
            "ks": dkLen*8,
            "ct": base64.b64encode(ciphertext),
            "iv": base64.b64encode(iv),
            "cipher": "aes",
            "mode": mode,
            "adata": "",
            "v": 1,
            "ts": tlen
            }
